fix(web): asset selection on memories page is broken

knechtandreas opened the pull request:
## Description

This fixes a couple of bugs around asset interaction on the memories page. Mainly if you scroll down to the gallery at the bottom of a memory and make a selection, then hit 'right/left' arrows to move to another photo, the selection would be cleared at the bottom but, but still be highlighted at the top. If you select the same photo again, it would be counted twice.

The underlying problem was that the existing memory-store would often be re-initialized resulting in a re-load of assets. Then these were shoved in a secondary store used for display and given that `asset-interaction` tracks full mutable object references of assets, you could easily get into a state where the re-load of assets by this memory store would conflict with an older version of the same asset object ref and therefore be counted/selected again.

There was also quite a few more issues around mutating state (favoriting photos removing them etc), so I went a bit further with this fix:

* Moved all store logic into the store (including talking to the remote back-end and updating the client side state)
* The store now is initialized only once to avoid re-loading duplicate asset object refs
* The store now just has the memories loaded as 'state' and this is the only thing that's mutated when actions are taken (favoriting etc). 
* It then has a derived object that transforms these memories to make the view/display easier (just as before but it's now not a store and not mutated any longer).

Fixes #16712 

## How Has This Been Tested?

Did lots of manual testing in browser.

## API Changes

* /memories/* APIs now correctly return `isFavorite` flag. Previously it was always returning `false` due to not passing along the user auth information.


## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [ ] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

knechtandreas committed:
16712: Proper intialisation of the memory store to avoid loading up duplicate object refs of the same asset.

knechtandreas committed:
16712: Add auth to memory mapping so isFavorite is actually return correctly from the server.

knechtandreas committed:
16712: Move logic that belongs in the store into the store.

knechtandreas committed:
Merge branch 'main' into fix/16712-asset-selection-on-memories-page-is-broken

* main:
  feat(mobile): person age on photo properties (#16728)
  refactor: migrate library spec to factories (#16711)
  chore: remove unused file (#16707)
  fix(mobile): fix notification icon not displaying properly (#16710)

knechtandreas committed:
16712: Cleanup.

knechtandreas committed:
16712: Fix init behaviour.

knechtandreas commented on server/src/dtos/memory.dto.ts:
```diff
@@ -102,6 +103,6 @@ export const mapMemory = (entity: MemoryItem): MemoryResponseDto => {
     data: entity.data as unknown as MemoryData,
     isSaved: entity.isSaved,
-    assets: ('assets' in entity ? entity.assets : []).map((asset) => mapAsset(asset as AssetEntity)),
+    assets: ('assets' in entity ? entity.assets : []).map((asset) => mapAsset(asset as AssetEntity, { auth })),
```
Because `auth` was missing here, `isFavorite` was always returning as `false` in the returned assets inside a memory. As a result the gallery view wouldn't show them as favorites even after just performing that action on the memory view.

knechtandreas commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -202,97 +152,59 @@
-  };
 
-  const handleRemove = (ids: string[]) => {
+  const handleDeleteOrArchiveAssets = (ids: string[]) => {
```
Re-named some of these methods to make it a bit easier to understand what they are used for. They all follow the same pattern now too:
* Store does the actualy mutation of state (both on server & client)
* Any display logic (e.g. notificationController) remains here.

knechtandreas commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -202,97 +152,59 @@
+  };
+
+  onMount(async () => {
+    await memoryStore.initialize();
```
This is the bugfix for the main bug (duplicate asset object refs for same assets). Now this initialize() method internally ensures we only call it once. Previously `loadMemories` was called from a couple of places, not always guarded to avoid a reload.

knechtandreas commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -350,24 +262,24 @@
         <AddToAlbum shared />
       </ButtonContextMenu>
 
-      <FavoriteAction removeFavorite={assetInteraction.isAllFavorite} onFavorite={handleUpdate} />
```
handleUpdate wasn't actually doing anything.

knechtandreas commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -350,24 +262,24 @@
 {/if}
 
 <section id="memory-viewer" class="w-full bg-immich-dark-gray" bind:this={memoryWrapper}>
-  {#if current && current.memory.assets.length > 0}
```
We only ever display memories that more than 0 assets. So I internalized this check in the memory store, which means it's no longer needed here in the view layer anywhere.

knechtandreas commented on web/src/lib/stores/memory.store.svelte.ts:
```diff
@@ -0,0 +1,117 @@
+  memories = $state<MemoryResponseDto[]>([]);
+  private initialized = false;
+  private memoryAssets = $derived.by(() => {
+    const memoryAssets: MemoryAsset[] = [];
```
This code came straight from the view layer. Except it's just a `$derived` state here now and it's not mutated anylonger.

knechtandreas commented on web/src/lib/stores/memory.store.svelte.ts:
```diff
@@ -0,0 +1,117 @@
+    return this.memoryAssets.find((memoryAsset) => memoryAsset.asset.id === assetId) ?? this.memoryAssets[0];
+  }
+
+  hideAssetsFromMemory(ids: string[]) {
```
Added clean API in the store here for any actions that might change state. These methods all follow same pattern:
* Make call to server first (if required)
* If that succeeded, only mutate state in `memories`

knechtandreas committed:
16712: Add comment.

knechtandreas committed:
16712: Make method private.

knechtandreas committed:
16712: Fix import.

knechtandreas committed:
16712: Fix format.

knechtandreas committed:
16712: Cleaner if/else and fix typo.

knechtandreas committed:
Merge branch 'main' into fix/16712-asset-selection-on-memories-page-is-broken

* main:
  fix(server): set the dev server restart policy of the dev server container to match the other containers (#16753)
  feat(web): remember search context (#16614)
  feat(server): read Android and Sony video camera make/model (#16678)
  fix(server): adjust type of person.birthDate (#16628)
  fix(web): add labels to memory lane buttons (#16664)
  feat(mobile): locate in timeline (#16722)
  chore(ml): uv (#16725)
  fix: 🍪 packages confusion (#16735)
  fix(web): Update people-card favorite position (#16746)
  chore(mobile): upgrade riverpod (#16742)
  chore(mobile): upgrade flutter_web_auth_2 (#16741)
  fix(docs): edge case when restoring dump that is unreadable as current user (#16758)

alextran1502 reviewed (approved):
Thank you! your PRs are always high quality and impressive.

alextran1502 committed:
fix: icon size mismatch

alextran1502 commented:
I've found a bug in the PR.

If you have more than two memories in the memory lane, choosing the second memory onward, the slideshow won't play automatically. The first memory always plays succesfully

knechtandreas commented:
> I've found a bug in the PR.
> 
> If you have more than two memories in the memory lane, choosing the second memory onward, the slideshow won't play automatically. The first memory always plays successfully
> 
> Here is the log from the browser's console
> 
> <img alt="image" width="570" src="https://private-user-images.githubusercontent.com/27055614/421544293-6fb09e5f-6d8c-4654-a858-6aeaf31e60b0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3MjYwNjEsIm5iZiI6MTc0MTcyNTc2MSwicGF0aCI6Ii8yNzA1NTYxNC80MjE1NDQyOTMtNmZiMDllNWYtNmQ4Yy00NjU0LWE4NTgtNmFlYWYzMWU2MGIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzExVDIwNDI0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkYWRmMDRlMTgxMmMzODU4OTFjMmZmOTZiMzhlYjExMzA2NTAzYjYxNGU2NDE5OGMxOTg3NGE5NzMzYWZiOTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.n0XuGi699Y1UToVVo3K5z858AeY4820pnKKwwiavSII">

Huh - sorry about that! I'm away today, but will take a look into this tomorrow.

knechtandreas committed:
16712: Fixed up state machine managing memory playback:
* Updated to `Tween` (`tweened` was deprecated)
* Removed `resetPromise`. Setting progressController to 0 had the same effect, so not really sure why it was there?
* Removed the many duplicate places the `handleAction` method was called. Now we just called it on `afterNavigate` as well as when `galleryInView` or `$isViewing` state changes.

knechtandreas commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -66,20 +66,15 @@
-  let progressBarController = tweened<number>(0, {
-    duration: (from: number, to: number) => (to ? 5000 * (to - from) : 0),
-  });
+  let progressBarController: Tween<number> | undefined = $state(undefined);
```
Updated to non-deprecated `Tween` API here, and initialising it to `undefined`. How to initialise this really depends on the type of asset (video / non-video) and this initialisation was immediately being override anyways.

knechtandreas commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -66,20 +66,15 @@
   let paused = $state(false);
   let current = $state<MemoryAsset | undefined>(undefined);
   let isSaved = $derived(current?.memory.isSaved);
-  let resetPromise = $state(Promise.resolve());
```
I'm not really sure why this existed. Just setting the progressBarController to 0 has exactly the same effect and simplifies the logic below.

knechtandreas commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -129,28 +127,49 @@
     }
   };
   const handleProgress = async (progress: number) => {
+    if (!progressBarController) {
```
This is probably superfluous (now that we only call `handleAction` from a few places it's initialised already).

knechtandreas commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -129,28 +127,49 @@
+        await progressBarController.set(0);
+        await handleNextAsset();
+      } else {
+        await handleAction('pause');
```
Now when we're on the very last asset we just pause and keep the progress bar full. Shows to the user we've finished. Previously it would always empty out the progressbar when playback had finished.

knechtandreas commented:
> I've found a bug in the PR.
> 
> If you have more than two memories in the memory lane, choosing the second memory onward, the slideshow won't play automatically. The first memory always plays successfully
> 
> Here is the log from the browser's console
> 
> <img alt="image" width="570" src="https://private-user-images.githubusercontent.com/27055614/421544293-6fb09e5f-6d8c-4654-a858-6aeaf31e60b0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE3NTU5OTMsIm5iZiI6MTc0MTc1NTY5MywicGF0aCI6Ii8yNzA1NTYxNC80MjE1NDQyOTMtNmZiMDllNWYtNmQ4Yy00NjU0LWE4NTgtNmFlYWYzMWU2MGIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEyVDA1MDEzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkMTczZGI1NDg1ZWQ3NWRlOWJiNjhhNWQxYmIwZTNlOWFmYTFkOWZjOWFhYTlkZmQwNGE4MjdkZThlY2JjNGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GuuaP7irVlr0eD3-qA-xzyDLZLNADISmPNQaXJfQKDo">

So while looking into the fix for this, I went down a rabbit hole a bit.  Fix is in this commit: https://github.com/immich-app/immich/pull/16759/commits/bdf729da283155e96069a7a2020756d0b299adac

I'm not 100% sure how this has worked reliably up until this point. In the existing code, the `handleAction` method that controls playback was called multiple times on first load of the memory-viewer component. Several of the initial calls were before some of the state it depends on was fully loaded yet. 

 Now I'm basically just calling it once at the end of `afterNavigate`, at which point everything it needs should be loaded up.

knechtandreas committed:
16712: Add aria tag.

knechtandreas commented:
Also,  apologies - for a 'simple' bugfix this PR got very messy. Should have perhaps tried to split up some of the re-factors/cleanups into separate PRs.

alextran1502 commented:
> Also, apologies - for a 'simple' bugfix this PR got very messy. Should have perhaps tried to split up some of the re-factors/cleanups into separate PRs.

No problem. :D I found an edge case that if the first memory is a video, the auto-play doesn't start. Here is the browser's log

<img width="586" alt="image" src="https://github.com/user-attachments/assets/5ba89aee-f852-41ef-a843-f7d0a301613a" />

knechtandreas commented:
> > Also, apologies - for a 'simple' bugfix this PR got very messy. Should have perhaps tried to split up some of the re-factors/cleanups into separate PRs.
> 
> No problem. :D I found an edge case that if the first memory is a video, the auto-play doesn't start. Here is the browser's log
> 
> <img alt="image" width="586" src="https://private-user-images.githubusercontent.com/27055614/421961936-5ba89aee-f852-41ef-a843-f7d0a301613a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE4MTcyNzUsIm5iZiI6MTc0MTgxNjk3NSwicGF0aCI6Ii8yNzA1NTYxNC80MjE5NjE5MzYtNWJhODlhZWUtZjg1Mi00MWVmLWE4NDMtZjdkMGEzMDE2MTNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEyVDIyMDI1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliOGIyNzcxOWQxNDc0MTQ1NjM5ZWY4YjI1OTI3Zjk1ZjYzNDYzMGVkODgxNTdjNGUyYjc1YzFiNWRjMDA1YmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.L3FdL1ydxbLPpgXhIomziM7zx9RpYsPoK_BoPTkzZPE">

Damn - thanks for finding that. Will fix.  Just did some testing on latest production code (without my changes) and even though the video plays when it's first in the memory, I also get some of those same console errors.

Will look at a proper fix.

knechtandreas committed:
16712: Fix memory player duplicate invocation bugs. Now we should only call 'reset' and 'play' once, after navigate/page load. This should hopefully fix all the various bugs around playback.

knechtandreas committed:
Merge branch 'main' into fix/16712-asset-selection-on-memories-page-is-broken

* main: (41 commits)
  chore(deps): update typescript-projects (#16795)
  fix: immich ui toggles and switches (#16834)
  feat: better library rename UX (#16837)
  refactor: use immich/ui button component in user settings (#16836)
  chore(mobile): bump dependency versions (#16823)
  fix(web): fixed formatting of video length (#16829)
  refactor: user entity (#16655)
  fix(web): update search results when searching from info panel (#16729)
  fix(docs): logo not loading dark theme variant in production (#16820)
  refactor: use factory and kysely types for partner repository (#16812)
  fix(ml): dev environment dependencies (#16815)
  fix: run preview label remove job on PR close (#16811)
  feat(web): Add keyboard shortcut selection on grid (#16713)
  feat(web): better person naming interface (#16631)
  chore(mobile): use path provider foundation (#16804)
  chore(mobile): add orientation tests for exif (#16806)
  chore: ignore correct build folder (#16808)
  fix(server): set unit test timezone to UTC (#16805)
  chore: shared suffix for docker tags (#16727)
  chore(deps): update base-image to v20250311 (major) (#16803)
  ...

knechtandreas committed:
16712: Cleanup

knechtandreas committed:
16712: Cleanup

knechtandreas committed:
16712: Cleanup

knechtandreas committed:
16712: Cleanup

knechtandreas commented:
Ok, so I really hope it's fixed now.  Did a lot of testing with/without videos and x-browser.  Tried loading memories with:
* Click on memory from photos page
* Navigate direct via URL
* Reload page
* Navigate by clicking between memories
* Keyboard shortcuts.

There were a lot of problems/bugs here (also on master), mainly due to the fact that we were calling `handleAction` many times in very unpredictable ways and order. Plus, `handleAction` clearly isn't idempotent at the moment. Here are just some ways `handleAction` was being invoked multiple times on page load and otherwise:
* onMount (usually after afterNavigate)
* afterNavigate
* sometimes when progress was being updated and reached certain states
* in various `$effect` handlers depending on if gallery or viewer was shown.

There was also an additional problem where we tried to call `play()` on the video element before it was mounted in the DOM (sometimes - timing dependant in my testing).

I've cleaned this up now by:
* Removing `onMount`. Instead, all instantiation is handled in `afterNavigate`
* Added some guards to ensure we only initialize once when a new asset is shown
* Removed the $effect handlers/fixed them up so they only invoke `handleAction` when necessary.

There's still one way video playback can fail, and that's if the browser blocks it.  E.g. autoplayback is disabled by the user, videos wont auto-play. Or some browsers will not play videos on page load if there's no user interaction with the DOM first. That's not fixable, but at least we handle it gracefully now by just pausing playback (so the user can then hit play).

I think  once this has been merged in, I might take another look at this code to see if we can use a state-machine or something similar to clean this up further and make it testable. I fear it's still going to be easy to introduce bugs at the moment.

knechtandreas commented:
I've created a spike PR now to show what this code could look like if we used a state machine (provided by [Xstate](https://stately.ai/docs/xstate)).  I think that approach is much cleaner and will be more robust/bugfree going forward, but I'm more than happy for it to be rejected as well if people think it's too heavyweight to bring in this library for this use case.

I created a branch/PR for the spike in my repo here for people to take a look at: https://github.com/knechtandreas/immich/pull/1

Will post to Discord #contributing as well.

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into fix/16712-asset-selection-on-memories-page-is-broken
